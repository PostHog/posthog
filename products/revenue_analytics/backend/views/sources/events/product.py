from collections.abc import Iterable

from posthog.hogql import ast

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_event
from products.revenue_analytics.backend.views.sources.helpers import events_expr_for_team


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    team = handle.team
    for event in team.revenue_analytics_config.events:
        if not event.productProperty:
            continue

        prefix = view_prefix_for_event(event.eventName)

        events_query = ast.SelectQuery(
            distinct=True,
            select=[
                ast.Alias(alias="product_id", expr=ast.Field(chain=["events", "properties", event.productProperty]))
            ],
            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
            where=events_expr_for_team(team),
        )

        query = ast.SelectQuery(
            select=[
                ast.Alias(alias="id", expr=ast.Field(chain=["product_id"])),
                ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
                ast.Alias(alias="name", expr=ast.Field(chain=["product_id"])),
            ],
            select_from=ast.JoinExpr(table=events_query),
            order_by=[ast.OrderExpr(expr=ast.Field(chain=["id"]), order="ASC")],
        )

        yield BuiltQuery(key=event.eventName, prefix=prefix, query=query)
