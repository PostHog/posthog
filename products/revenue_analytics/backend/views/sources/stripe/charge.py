from typing import cast
from collections.abc import Iterable

from posthog.hogql import ast
from posthog.temporal.data_imports.sources.stripe.constants import (
    CHARGE_RESOURCE_NAME as STRIPE_CHARGE_RESOURCE_NAME,
)
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable
from posthog.hogql.database.schema.exchange_rate import (
    EXCHANGE_RATE_DECIMAL_PRECISION,
    convert_currency_call,
)
from products.revenue_analytics.backend.views.sources.helpers import (
    is_zero_decimal_in_stripe,
    currency_aware_divider,
    currency_aware_amount,
)

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_source


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    source = handle.source
    if source is None:
        return

    # Get all schemas for the source, avoid calling `filter` and do the filtering on Python-land
    # to avoid n+1 queries
    schemas = source.schemas.all()
    charge_schema = next((schema for schema in schemas if schema.name == STRIPE_CHARGE_RESOURCE_NAME), None)
    if charge_schema is None:
        return

    charge_schema = cast(ExternalDataSchema, charge_schema)
    if charge_schema.table is None:
        return

    table = cast(DataWarehouseTable, charge_schema.table)
    team = table.team

    prefix = view_prefix_for_source(source)

    query = ast.SelectQuery(
        select=[
            # Base fields to allow insights to work (need `distinct_id` AND `timestamp` fields)
            ast.Alias(alias="id", expr=ast.Field(chain=["id"])),
            ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
            ast.Alias(alias="timestamp", expr=ast.Field(chain=["created_at"])),
            # Useful for cross joins
            ast.Alias(alias="customer_id", expr=ast.Field(chain=["customer_id"])),
            ast.Alias(alias="invoice_id", expr=ast.Field(chain=["invoice_id"])),
            # Empty, but required for the `events` view to work
            ast.Alias(alias="session_id", expr=ast.Constant(value=None)),
            ast.Alias(alias="event_name", expr=ast.Constant(value=None)),
            # Compute the original currency, converting to uppercase to match the currency code in the `exchange_rate` table
            ast.Alias(
                alias="original_currency",
                expr=ast.Call(name="upper", args=[ast.Field(chain=["currency"])]),
            ),
            # Compute the original amount in the original currency
            # by looking at the captured amount, effectively ignoring refunded value
            ast.Alias(
                alias="original_amount",
                expr=ast.Call(
                    name="toDecimal",
                    args=[
                        ast.Field(chain=["amount_captured"]),
                        ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION),
                    ],
                ),
            ),
            # Compute whether the original currency is a zero-decimal currency
            # by comparing it against a list of known zero-decimal currencies
            # in Stripe's API
            ast.Alias(
                alias="enable_currency_aware_divider",
                expr=is_zero_decimal_in_stripe(ast.Field(chain=["original_currency"])),
            ),
            # Compute the amount decimal divider, which is 1 for zero-decimal currencies and 100 for others
            # This is used to convert the original amount to the adjusted amount
            currency_aware_divider(),
            # Compute the adjusted original amount, which is the original amount divided by the amount decimal divider
            currency_aware_amount(),
            # Expose the base/converted currency, which is the base currency from the team's revenue config
            ast.Alias(alias="currency", expr=ast.Constant(value=team.base_currency)),
            # Convert the adjusted original amount to the base currency
            ast.Alias(
                alias="amount",
                expr=convert_currency_call(
                    amount=ast.Field(chain=["currency_aware_amount"]),
                    currency_from=ast.Field(chain=["original_currency"]),
                    currency_to=ast.Field(chain=["currency"]),
                    timestamp=ast.Call(
                        name="_toDate",
                        args=[
                            ast.Call(
                                name="ifNull",
                                args=[
                                    ast.Field(chain=["timestamp"]),
                                    ast.Call(name="toDateTime", args=[ast.Constant(value=0)]),
                                ],
                            ),
                        ],
                    ),
                ),
            ),
        ],
        # Simple query, simply refer to the `stripe_charge` table
        select_from=ast.JoinExpr(table=ast.Field(chain=[table.name])),
        # Only include succeeded charges because they're the ones that represent revenue
        where=ast.CompareOperation(
            left=ast.Field(chain=["status"]),
            right=ast.Constant(value="succeeded"),
            op=ast.CompareOperationOp.Eq,
        ),
    )

    yield BuiltQuery(key=str(table.id), prefix=prefix, query=query)
