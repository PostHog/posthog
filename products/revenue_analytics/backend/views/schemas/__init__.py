from posthog.schema import DatabaseSchemaManagedViewTableKind
from ._definitions import Schema

from .subscription import SCHEMA as SUBSCRIPTION
from .charge import SCHEMA as CHARGE
from .customer import SCHEMA as CUSTOMER
from .product import SCHEMA as PRODUCT
from .revenue_item import SCHEMA as REVENUE_ITEM

SCHEMAS: dict[DatabaseSchemaManagedViewTableKind, Schema] = {
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_SUBSCRIPTION: SUBSCRIPTION,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_REVENUE_ITEM: REVENUE_ITEM,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CHARGE: CHARGE,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CUSTOMER: CUSTOMER,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_PRODUCT: PRODUCT,
}
