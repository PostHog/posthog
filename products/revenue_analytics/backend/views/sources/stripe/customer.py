from typing import cast
from collections.abc import Iterable

from posthog.hogql import ast

from posthog.temporal.data_imports.sources.stripe.constants import (
    CUSTOMER_RESOURCE_NAME as STRIPE_CUSTOMER_RESOURCE_NAME,
    INVOICE_RESOURCE_NAME as STRIPE_INVOICE_RESOURCE_NAME,
)
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable
from products.revenue_analytics.backend.views.sources.helpers import (
    extract_json_string,
    get_cohort_expr,
)

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_source


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    source = handle.source
    if source is None:
        return

    # Get all schemas for the source, avoid calling `filter` and do the filtering on Python-land
    # to avoid n+1 queries
    schemas = source.schemas.all()
    customer_schema = next((schema for schema in schemas if schema.name == STRIPE_CUSTOMER_RESOURCE_NAME), None)
    if customer_schema is None:
        return

    customer_schema = cast(ExternalDataSchema, customer_schema)
    if customer_schema.table is None:
        return

    invoice_schema = next((schema for schema in schemas if schema.name == STRIPE_INVOICE_RESOURCE_NAME), None)
    invoice_table = None
    if invoice_schema is not None:
        invoice_schema = cast(ExternalDataSchema, invoice_schema)
        invoice_table = invoice_schema.table
        if invoice_table is not None:
            invoice_table = cast(DataWarehouseTable, invoice_table)

    table = cast(DataWarehouseTable, customer_schema.table)
    prefix = view_prefix_for_source(source)

    query = ast.SelectQuery(
        select=[
            ast.Alias(alias="id", expr=ast.Field(chain=["outer", "id"])),
            ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
            ast.Alias(alias="timestamp", expr=ast.Field(chain=["created_at"])),
            ast.Alias(alias="name", expr=ast.Field(chain=["name"])),
            ast.Alias(alias="email", expr=ast.Field(chain=["email"])),
            ast.Alias(alias="phone", expr=ast.Field(chain=["phone"])),
            ast.Alias(alias="address", expr=ast.Field(chain=["address"])),
            ast.Alias(alias="metadata", expr=ast.Field(chain=["metadata"])),
            ast.Alias(
                alias="country",
                expr=extract_json_string("address", "country"),
            ),
            ast.Alias(alias="cohort", expr=ast.Constant(value=None)),
            ast.Alias(alias="initial_coupon", expr=ast.Constant(value=None)),
            ast.Alias(alias="initial_coupon_id", expr=ast.Constant(value=None)),
        ],
        select_from=ast.JoinExpr(
            alias="outer",
            table=ast.Field(chain=[table.name]),
        ),
    )

    # If there's an invoice table we can generate the cohort entry
    # by looking at the first invoice for each customer
    if invoice_table is not None:
        cohort_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "cohort"), None
        )
        if cohort_alias is not None:
            cohort_alias.expr = ast.Field(chain=["cohort"])

        initial_coupon_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "initial_coupon"),
            None,
        )
        if initial_coupon_alias is not None:
            initial_coupon_alias.expr = ast.Field(chain=["initial_coupon"])

        initial_coupon_id_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "initial_coupon_id"),
            None,
        )
        if initial_coupon_id_alias is not None:
            initial_coupon_id_alias.expr = ast.Field(chain=["initial_coupon_id"])

        if query.select_from is not None and (
            cohort_alias is not None or initial_coupon_alias is not None or initial_coupon_id_alias is not None
        ):
            query.select_from.next_join = ast.JoinExpr(
                alias="cohort_inner",
                table=ast.SelectQuery(
                    select=[
                        ast.Field(chain=["customer_id"]),
                        ast.Alias(alias="cohort", expr=get_cohort_expr("min(created_at)")),
                        ast.Alias(
                            alias="initial_coupon",
                            expr=ast.Call(
                                name="argMin",
                                args=[
                                    extract_json_string("discount", "coupon", "name"),
                                    ast.Field(chain=["created_at"]),
                                ],
                            ),
                        ),
                        ast.Alias(
                            alias="initial_coupon_id",
                            expr=ast.Call(
                                name="argMin",
                                args=[
                                    extract_json_string("discount", "coupon", "id"),
                                    ast.Field(chain=["created_at"]),
                                ],
                            ),
                        ),
                    ],
                    select_from=ast.JoinExpr(alias="invoice", table=ast.Field(chain=[invoice_table.name])),
                    group_by=[ast.Field(chain=["customer_id"])],
                ),
                join_type="LEFT JOIN",
                constraint=ast.JoinConstraint(
                    constraint_type="ON",
                    expr=ast.CompareOperation(
                        left=ast.Field(chain=["cohort_inner", "customer_id"]),
                        right=ast.Field(chain=["outer", "id"]),
                        op=ast.CompareOperationOp.Eq,
                    ),
                ),
            )

    yield BuiltQuery(key=str(table.id), prefix=prefix, query=query)
