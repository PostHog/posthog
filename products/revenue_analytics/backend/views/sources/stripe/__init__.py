from posthog.schema import DatabaseSchemaManagedViewTableKind
from products.revenue_analytics.backend.views.core import Builder

from .charge import build as charge_builder
from .customer import build as customer_builder
from .product import build as product_builder
from .revenue_item import build as revenue_item_builder
from .subscription import build as subscription_builder

BUILDER: Builder = {
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CHARGE: charge_builder,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CUSTOMER: customer_builder,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_PRODUCT: product_builder,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_REVENUE_ITEM: revenue_item_builder,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_SUBSCRIPTION: subscription_builder,
}
