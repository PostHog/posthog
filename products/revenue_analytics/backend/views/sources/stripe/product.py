from typing import cast
from collections.abc import Iterable

from posthog.hogql import ast

from posthog.temporal.data_imports.sources.stripe.constants import (
    PRODUCT_RESOURCE_NAME as STRIPE_PRODUCT_RESOURCE_NAME,
)
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_source


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    source = handle.source
    if source is None:
        return

    # Get all schemas for the source, avoid calling `filter` and do the filtering on Python-land
    # to avoid n+1 queries
    schemas = source.schemas.all()
    product_schema = next((schema for schema in schemas if schema.name == STRIPE_PRODUCT_RESOURCE_NAME), None)
    if product_schema is None:
        return

    product_schema = cast(ExternalDataSchema, product_schema)
    if product_schema.table is None:
        return

    table = cast(DataWarehouseTable, product_schema.table)
    prefix = view_prefix_for_source(source)

    query = ast.SelectQuery(
        select=[
            ast.Alias(alias="id", expr=ast.Field(chain=["id"])),
            ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
            ast.Alias(alias="name", expr=ast.Field(chain=["name"])),
        ],
        select_from=ast.JoinExpr(table=ast.Field(chain=[table.name])),
    )

    yield BuiltQuery(key=str(table.id), prefix=prefix, query=query)
