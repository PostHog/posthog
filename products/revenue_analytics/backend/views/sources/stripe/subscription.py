from typing import cast
from collections.abc import Iterable

from posthog.hogql import ast
from posthog.temporal.data_imports.sources.stripe.constants import (
    SUBSCRIPTION_RESOURCE_NAME as STRIPE_SUBSCRIPTION_RESOURCE_NAME,
)
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_source
from products.revenue_analytics.backend.views.sources.helpers import extract_json_string


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    source = handle.source
    if source is None:
        return

    # Get all schemas for the source, avoid calling `filter` and do the filtering on Python-land
    # to avoid n+1 queries
    schemas = source.schemas.all()
    subscription_schema = next((schema for schema in schemas if schema.name == STRIPE_SUBSCRIPTION_RESOURCE_NAME), None)
    if subscription_schema is None:
        return

    subscription_schema = cast(ExternalDataSchema, subscription_schema)
    if subscription_schema.table is None:
        return

    table = cast(DataWarehouseTable, subscription_schema.table)
    prefix = view_prefix_for_source(source)

    # Even though we need a string query for the view,
    # using an ast allows us to comment what each field means, and
    # avoid manual interpolation of constants, leaving that to the HogQL printer
    #
    # These are all pretty basic, they're simply here to allow future extensions
    # once we start adding fields from sources other than Stripe
    query = ast.SelectQuery(
        select=[
            ast.Alias(alias="id", expr=ast.Field(chain=["id"])),
            ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
            ast.Alias(alias="plan_id", expr=extract_json_string("plan", "id")),
            ast.Alias(alias="product_id", expr=extract_json_string("plan", "product")),
            ast.Alias(alias="customer_id", expr=ast.Field(chain=["customer_id"])),
            ast.Alias(alias="status", expr=ast.Field(chain=["status"])),
            ast.Alias(alias="started_at", expr=ast.Field(chain=["created_at"])),
            # If has an end date, but it's in the future, then just not include `ended_at`
            ast.Alias(
                alias="ended_at",
                expr=ast.Call(
                    name="if",
                    args=[
                        ast.CompareOperation(
                            op=ast.CompareOperationOp.Gt,
                            left=ast.Field(chain=["ended_at"]),
                            right=ast.Call(name="today", args=[]),
                        ),
                        ast.Constant(value=None),
                        ast.Field(chain=["ended_at"]),
                    ],
                ),
            ),
            ast.Alias(alias="metadata", expr=ast.Field(chain=["metadata"])),
        ],
        select_from=ast.JoinExpr(table=ast.Field(chain=[table.name])),
    )

    yield BuiltQuery(key=str(table.id), prefix=prefix, query=query)
