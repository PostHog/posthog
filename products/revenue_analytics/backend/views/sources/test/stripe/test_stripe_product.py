from posthog.temporal.data_imports.sources.stripe.constants import PRODUCT_RESOURCE_NAME

from products.revenue_analytics.backend.views.sources.stripe.product import build
from products.revenue_analytics.backend.views.sources.test.stripe.base import StripeSourceBaseTest
from products.revenue_analytics.backend.views.schemas.product import SCHEMA as PRODUCT_SCHEMA


class TestProductStripeBuilder(StripeSourceBaseTest):
    def setUp(self):
        super().setUp()
        self.setup_stripe_external_data_source()

    def test_build_product_query_with_product_schema(self):
        """Test building product query when product schema exists."""
        # Setup with only product schema
        self.setup_stripe_external_data_source(schemas=[PRODUCT_RESOURCE_NAME])

        queries = list(build(self.stripe_handle))

        # Should build one query for the product schema
        self.assertEqual(len(queries), 1)

        product_query = queries[0]
        product_table = self.get_stripe_table_by_schema_name(PRODUCT_RESOURCE_NAME)

        # Test the query structure
        self.assertQueryContainsFields(product_query.query, PRODUCT_SCHEMA)
        self.assertBuiltQueryStructure(
            product_query, str(product_table.id), f"stripe.{self.external_data_source.prefix}"
        )

        # Print and snapshot the generated HogQL query
        query_sql = product_query.query.to_hogql()
        self.assertQueryMatchesSnapshot(query_sql, replace_all_numbers=True)

    def test_build_with_no_product_schema(self):
        """Test that build returns empty when no product schema exists."""
        # Setup without product schema
        self.setup_stripe_external_data_source(schemas=[])

        queries = list(build(self.stripe_handle))

        # Should return no queries
        self.assertEqual(len(queries), 0)

    def test_build_with_product_schema_but_no_table(self):
        """Test that build returns empty when product schema exists but has no table."""
        # Setup with product schema but no table
        self.setup_stripe_external_data_source_with_specific_schemas(
            [{"name": PRODUCT_RESOURCE_NAME, "table_name": None}]
        )

        # Set the table to None to simulate missing table
        product_schema = self.get_stripe_schema_by_name(PRODUCT_RESOURCE_NAME)
        product_schema.table = None

        queries = list(build(self.stripe_handle))

        # Should return no queries
        self.assertEqual(len(queries), 0)

    def test_build_with_no_source(self):
        """Test that build returns empty when source is None."""
        handle = self.create_stripe_handle_without_source()

        queries = list(build(handle))

        # Should return no queries
        self.assertEqual(len(queries), 0)

    def test_product_query_contains_required_fields(self):
        """Test that the generated query contains all required product fields."""
        self.setup_stripe_external_data_source(schemas=[PRODUCT_RESOURCE_NAME])

        queries = list(build(self.stripe_handle))
        product_query = queries[0]

        query_sql = product_query.query.to_hogql()

        # Check for specific fields in the query
        self.assertIn("id", query_sql)
        self.assertIn("source_label", query_sql)
        self.assertIn("name", query_sql)

        # Check that source_label contains the expected prefix
        expected_prefix = f"stripe.{self.external_data_source.prefix}"
        self.assertIn(f"'{expected_prefix}'", query_sql)
