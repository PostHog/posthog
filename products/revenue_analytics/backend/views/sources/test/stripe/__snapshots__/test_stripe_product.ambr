# serializer version: 1
# name: TestProductStripeBuilder.test_build_product_query_with_product_schema
  '''
  SELECT id AS id,
         'stripe.stripe_test' AS source_label,
         name AS name
  FROM stripe_test_product
  '''
# ---
